smolyay/surrogate.py

```diff
@@ -0,0 +1,279 @@
+        point_num = 0
+        for points in self._grids.points:
+            for dimension_ in range(self._dimension):
+                grid_num = 0
+                for basis_function in list(zip(
+                        *self._grids.basis_functions))[dimension_]:
+                    self._basis_matrix[point_num][grid_num] *= (
+                        basis_function(points[dimension_]))
+                    grid_num += 1
+            point_num += 1
+
+    def train(self, real_function, linear_solver='lu'):
+        """Fit surrogate's components (basis functions) to the function.
+
+        Parameters
+        ----------
+        real_function: callabe
+            Function to be approximated.
+        linear_slover: string
+            Method of solving linear equations:
+                'lu': lower upper decomposition
+                'inv': solve for coefficients through x = A^{-1}B.
+                'lstsq': least-square solution.
+
+        Returns
+        -------
+        list
+            Coefficients of the surrogate.
+        """
+        self.real_function = real_function
+        self.linear_solver = linear_solver
+        self._make_basis_matrix()
+
+        # transform grid points
+        transformed_grids = numpy.zeros((
+            len(self._grids.points), self._dimension))
+        for dimension_ in range(self._dimension):
+            transformed_grids[:, dimension_] = (
+                numpy.polynomial.polyutils.mapdomain(
+                    numpy.array(self._grids.points)[:, dimension_],
+                    (-1, 1), self.domain[dimension_]))
+
+        # evaluate real function at grid points
+        real_function_at_grids = numpy.zeros((len(self._grids.points)))
+        for point_num in range(len(self._grids.points)):
+            real_function_at_grids[point_num] = (
+                self.real_function(*transformed_grids[point_num]))
+
+        # fit
+        self._coefficients = solve_linear(
+            self._basis_matrix, real_function_at_grids,
+            self.linear_solver)
+
+        return self._coefficients.tolist()
+
+    def train_from_data(self, real_function_at_grids, linear_solver='lu'):
+        """Fit surrogate's components (basis functions) to data.
+
+        Parameters
+        ----------
+        real_function_at_grids: list
+            Real function at grid points.
+        linear_slover: string
+            Method of solving linear equations:
+                'lu': lower upper decomposition
+                'inv': solve for coefficients through x = A^{-1}B.
+                'lstsq': least-square solution.
+
+        Returns
+        -------
+        list
+            Coefficients of the surrogate.
+
+        Raises
+        ------
+        IndexError
+            Data should be of length of the grid points.
+        """
+        if len(real_function_at_grids) != len(self._grids.points):
+            raise IndexError("Data must be in length of the grid points.")
```

Comment:
You should force the data to be an `ndarray` with `ndmin=2`, and then check the shape
```suggestion
        if real_function_at_grids.shape != self._grids.points.shape:
            raise IndexError("Data must be in length of the grid points.")
```

Reply:
This wouldn't work because shape of the `grid_points` is (# of grid points, dimension). However, shape of the `real_function_at_grids` is (1, # of grid points). 
This however work: 
`numpy.array(real_function_at_grids, ndmin=2).shape[1] != (numpy.array(self.grid.points).shape[0])`